habitat_baselines/common/rollout_storage.py

```diff
@@ -4,237 +4,196 @@
-                self.returns[step] = gae + self.value_preds[step]
         else:
-            self.returns[self.step] = next_value
+            self.buffers["returns"][self.step] = next_value
             for step in reversed(range(self.step)):
-                self.returns[step] = (
-                    self.returns[step + 1] * gamma * self.masks[step + 1]
-                    + self.rewards[step]
+                self.buffers["returns"][step] = (
+                    gamma
+                    * self.buffers["returns"][step + 1]
+                    * self.buffers["masks"][step + 1]
+                    + self.buffers["rewards"][step]
                 )
 
-    def recurrent_generator(self, advantages, num_mini_batch):
-        num_processes = self.rewards.size(1)
-        assert num_processes >= num_mini_batch, (
-            "Trainer requires the number of processes ({}) "
+    def recurrent_generator(self, advantages, num_mini_batch) -> TensorDict:
+        num_simulators = advantages.size(1)
+        assert num_simulators >= num_mini_batch, (
+            "Trainer requires the number of simulators ({}) "
             "to be greater than or equal to the number of "
-            "trainer mini batches ({}).".format(num_processes, num_mini_batch)
+            "trainer mini batches ({}).".format(num_simulators, num_mini_batch)
         )
-        num_envs_per_batch = num_processes // num_mini_batch
-        perm = torch.randperm(num_processes)
-        for start_ind in range(0, num_processes, num_envs_per_batch):
-            observations_batch = defaultdict(list)
-
-            recurrent_hidden_states_batch = []
-            actions_batch = []
-            prev_actions_batch = []
-            value_preds_batch = []
-            return_batch = []
-            masks_batch = []
-            old_action_log_probs_batch = []
-            adv_targ = []
-
-            for offset in range(num_envs_per_batch):
-                ind = perm[start_ind + offset]
-
-                for sensor in self.observations:
-                    observations_batch[sensor].append(
-                        self.observations[sensor][: self.step, ind]
-                    )
-
-                recurrent_hidden_states_batch.append(
-                    self.recurrent_hidden_states[0, :, ind]
-                )
-
-                actions_batch.append(self.actions[: self.step, ind])
-                prev_actions_batch.append(self.prev_actions[: self.step, ind])
-                value_preds_batch.append(self.value_preds[: self.step, ind])
-                return_batch.append(self.returns[: self.step, ind])
-                masks_batch.append(self.masks[: self.step, ind])
-                old_action_log_probs_batch.append(
-                    self.action_log_probs[: self.step, ind]
+        if num_simulators % num_mini_batch != 0:
+            warnings.warn(
+                "Number of simulators ({}) is not a multiple of the number of mini batches ({})".format(
```

Comment:
Possibly, we can add why that `Number of simulators should be a multiple of the number of mini batches`